core/controllers/acl_decorators.py

```diff
@@ -3527,3 +3527,46 @@ def test_can_update_suggestion(
 
     test_can_update_suggestion.__wrapped__ = True
     return test_can_update_suggestion
+
+
+def is_from_oppia_android(handler):
+    """Decorator to check whether the request was sent from Oppia Android.
+
+    Args:
+        handler: function. The function to be decorated.
+
+    Returns:
+        function. The newly decorated function.
+    """
+
+    def test_is_from_oppia_android(self, **kwargs):
+        """Checks whether the request was sent from Oppia Android.
+
+        Args:
+            **kwargs: *. Keyword arguments.
+
+        Returns:
+            *. The return value of the decorated function.
+
+        Raises:
+            UnauthorizedUserException. If incoming request is not from a valid
+                Oppia Android request.
+        """
+        headers = self.request.headers
+        api_key = headers['api_key']
+        app_package_name = headers['app_package_name']
+        app_version_name = headers['app_version_name']
+        app_version_code = headers['app_version_code']
+        if (
+                api_key != feconf.ANDROID_API_KEY or
+                app_package_name != feconf.ANDROID_APP_PACKAGE_NAME or
+                app_version_name != feconf.ANDROID_APP_VERSION_NAME or
+                app_version_code != feconf.ANDROID_APP_VERSION_CODE):
```

Comment:
This seems potentially wrong. There will be multiple deployed apps with different version names & version codes--we don't actually want to disable traffic to those otherwise it would break all older app versions.

Reply:
Ack, updated this to validate against a list of valid codes & versions. Since these are passed as strings in the headers we couldn't specify a range of valid codes / version so we'll have to specify a list of allowed version, although I'm not sure this is the best solution either given the asynchronous release schedule. That said, we have to validate these headers somehow -- PTAL and LMK your thoughts.

Reply:
Could we take an approach similar to what platform parameters does? In particular, we can validate the version by making sure that it matches our naming scheme (the current list with "1.0" won't actually match since versions follow the scheme: _major.minor-flavor-commithash_. See: https://github.com/oppia/oppia/blob/develop/core/domain/platform_parameter_domain.py#L147.

It may be useful to extract the values used there for validating versions & moving them to android_constants to ensure all critical compatibility values are kept within the same constants file, and to use the same ones here to validate that version is correct.

For version code, can we just verify that it's bigger than 0 & an integer? Eventually, we can introduce more targeted range checks once we know what the GA release will be but we can't do that yet.

Reply:
Done -- added them to `android_validation_constants` but I did not remove them from the `constants` file since they're used in some other places and I don't want to expand this PR too much more.

Which of [these constants](https://github.com/oppia/oppia/blob/develop/assets/constants.ts#L5828-L5839) would be android-specific and worth moving to `android_validation_constants`? I can create an issue for that.

Reply:
The platform parameters ones are all probably worth moving, I think.

Reply:
Filed issue [13626](https://github.com/oppia/oppia/issues/13626)